azure/Guidelines.md

```diff
@@ -88,80 +86,53 @@ All Azure APIs **MUST** use explicit versioning. The Microsoft REST API guidelin
 
 The Microsoft REST API guidelines give two options for how services and clients communicate the version: a url segment and a query parameter. Azure services **MUST** use the api-version query parameter. For example:
 
-```
+```text
 GET https://blobstore.azure.com/foo.com/acct1/c1/blob2?api-version=1.0
 PUT https://blobstore.azure.com/foo.com/acct1/c1/b2?api-version=2014-12-07
 POST https://blobstore.azure.com/foo.com/acct1/c1/b2?api-version=2015-12-07
 ```
 
-### Breaking changes in Azure
-
-A breaking change is any change in the API that may cause client or service code making the API call to fail. Obvious examples of such a change are the removal of an endpoint, adding or removing a required field or changing the format of the body (from XML to JSON for example).
-
-Even though we recommend clients ignore new fields, there are many libraries and clients that fail when new fields are introduced. Azure services **MUST** update the version number of their API even when adding optional fields. In fact, servers should be as strict as possible. Ignoring a field can result in the API accepting content that containered a typo or an element at the wrong level of nesting. If this missing field changes the semantics (for example, we have seen cases where security settings were misplaced and ignored, leaving the resources more exposed than intended) this can be a huge and hard to discover error.
```

Comment:
I want to challenge this piece of accepted wisdom.   At the very least, if it's going to state libraries, it should cite some popular ones that behave like this.   This one statement drives so much decision making and given the experience we have in other groups at Microsoft, I simply don't buy it.  It may have been true at the dawn of Azure, but I don't believe it is true now.   And given the customer pain in having to track and chase version changes caused by this when the large majority of change is additive and wouldn't otherwise require a version change at all, there is a business decision to be made between pain for a tiny minority of customers vs pain for a large majority thereof.

Reply:
Trying to find some middle ground... what if we allowed clients to opt-in to picking up evolutionary changes without a version bump?

`GET https://blobstore.azure.com/foo.com/acct1/c1/blob2?api-version=1.x`

Resolves to the latest 1.x version of the API. It doesn't solve service side concerns but does allow customers to opt into less conservative behavior.

Reply:
That's pretty interesting - we are very used to doing ~ or ^ as version prefixes to semver in our NPM packages in order to say, hey I want to pick newer stuff up until there's a major change.

Reply:
And I am ALL for customers getting to opt-in to te level of sophistication they happen to think about as a client.